daemon/qrexec-daemon.c

```diff
@@ -1102,16 +1143,54 @@ static void handle_execute_service(
                      type,
                      target_domain) <= 0)
             _exit(126);
+        char *cid;
+        if (asprintf(&cid, "%s,%s,%d", request_id->ident, remote_domain_name,
+                     remote_domain_id) <= 0)
+            _exit(126);
+
+        const char *to_exec[] = {
+            "/usr/bin/qrexec-client",
+            "-Ed@adminvm",
+            "-c",
+            cid,
+            "--",
+            cmd,
+            NULL,
+        };
+        execv(to_exec[0], (char **)to_exec);
+        LOG(ERROR, "execve() failed: %m");
+        _exit(126);
     } else {
-        char *buf;
-        if (strncmp("@dispvm:", target, sizeof("@dispvm:") - 1) == 0) {
-            disposable = true;
-            buf = qubesd_call(target + 8, "admin.vm.CreateDisposable", "", &resp_len);
+        bool disposable = strncmp("@dispvm:", target, sizeof("@dispvm:") - 1) == 0;
+        bool use_uuid = target_uuid != NULL;
+        if (disposable) {
+            char *buf = qubesd_call2(use_uuid ? target_uuid : target + 8,
+                                     "admin.vm.CreateDisposable", "",
+                                     "uuid", use_uuid ? 4 : 0, &resp_len);
             if (!buf) // error already printed by qubesd_call
                 _exit(126);
             if (memcmp(buf, "0", 2) == 0) {
-                /* we exec later so memory leaks do not matter */
```

Comment:
This comment is gone, but it's still applicable.

Reply:
We only exec for VM ⇒ dom0 service calls.  VM ⇒ VM calls do not exec.